In [1]:
#Libraries
import pandas as pd
import numpy as np
import sklearn
import sqlite3

In [25]:
#Opening a sqlite3 connection, and preparing the query for panda.
#We will be using Telco's Customer Churn data for this project.
#The end goal: Reach a dataframe with mostly numeric datatype that a machine learning model can fit and predict.

conn = sqlite3.connect("churn.db")
query = "SELECT * FROM Customers"

In [24]:
#Read the .db file using .read_sql() method. The connection is closed as there are no further uses.

df = pd.read_sql(query + " WHERE Churn LIKE 'Yes'", conn, index_col = 'customerID')
conn.close()

In [15]:
#The .info() method.

df.info()

<class 'pandas.DataFrame'>
Index: 1869 entries, 3668-QPYBK to 8361-LTMKD
Data columns (total 20 columns):
 #   Column            Non-Null Count  Dtype
---  ------            --------------  -----
 0   gender            1869 non-null   str  
 1   SeniorCitizen     1869 non-null   str  
 2   Partner           1869 non-null   str  
 3   Dependents        1869 non-null   str  
 4   tenure            1869 non-null   str  
 5   PhoneService      1869 non-null   str  
 6   MultipleLines     1869 non-null   str  
 7   InternetService   1869 non-null   str  
 8   OnlineSecurity    1869 non-null   str  
 9   OnlineBackup      1869 non-null   str  
 10  DeviceProtection  1869 non-null   str  
 11  TechSupport       1869 non-null   str  
 12  StreamingTV       1869 non-null   str  
 13  StreamingMovies   1869 non-null   str  
 14  Contract          1869 non-null   str  
 15  PaperlessBilling  1869 non-null   str  
 16  PaymentMethod     1869 non-null   str  
 17  MonthlyCharges    1869 non-null   

The output of the cell directly above this indicates that there are no records that contains a single null value, or in other words, no missing values. The .info() provides technical information about a dataframe, bundled with .head() and .tail().

In [16]:
#Using the .shape attribute. Note that 'shape' is an attribute, not a method - we do not call 'shape' on the dataframe, but rather, we see what shape it has.

df.shape

(1869, 20)

As you can see, there are 1869 rows and 20 columns, excluding the index_col.

(!) Note: Regarding .info() from 2 cells above, I stated that there are no missing values, which is incorrect. I imported the .csv to sqlite using .import [filepath] Customers, which imports the .csv file into a table in a .db file. This method has a fatal flaw - .import reads everything as TEXT, so the .read_sql() method would return a table whose dtypes are all str, including ' ' - the blank string, present in TotalCharges. Therefore, the dataframe may have missing values, hiding as strings - they are still technically non-null values.

In [67]:
df["TotalCharges"] = pd.to_numeric(df["TotalCharges"])
df["MonthlyCharges"] = pd.to_numeric(df["MonthlyCharges"])
df[["MonthlyCharges", "TotalCharges"]].isna().sum()

#Note: df[] takes a single string, or a LIST of string as input, never two or more strings separted by a ','.

MonthlyCharges    0
TotalCharges      0
dtype: int64

Safe to say the two columns above have no missing values.